### ISSUE TITLE
Generalize Babel Plugin

### ISSUE THREAD

**Issue by devknoll** [facebook/relay#206]
It's kind of a bummer that folks need to write their own `babelRelayPlugin.js` for each project. I'm going to add support for a `.graphqlrc` file that points to a schema instead.

I heard there were plans for the plugin going forward. Can someone expand on those? Would this be something you'd like to see merged into the main plugin?

**Comment by yungsters** [comment]
I'm actually not familiar with these "plans for the plugin going forward". @josephsavona?

**Comment by devknoll** [comment]
> Going forward, we'll provide additional documentation and make it easier to work with this plugin. Expect more information soon.
- https://facebook.github.io/relay/docs/guides-babel-plugin.html#content

**Comment by josephsavona** [comment]
@devknoll The goal is to be able to create a plugin from a `schema.js` file directly, instead of having to convert from schema -> introspection JSON. The blocker for this is that the schema file ends up `require()`-ing a different instance of `graphql` than the plugin does, and `instanceof` checks fail in `GraphQLPrinter.js`.

**Comment by fson** [comment]
@josephsavona However that doesn't help people who have their GraphQL server in another repository, running on a different server or using one of the growing number of other server implementation than graphql-js.

I'm more interested in fetching the schema using the introspection query, since it can work in these scenarios as well. Generalizing it might have some difficulties, because of varying authentication methods people might use etc. But maybe we can still provide a script that, like DefaultNetworkLayer in Relay, works for simpler cases out of the box so people don't need to worry about this when they are just getting started?

**Comment by josephsavona** [comment]
@devknoll I was just commenting about making integration with JS schemas easier, but I agree that we should also simplify things for products using alternate GraphQL implementations.

**Comment by amccloud** [comment]
@fson i'm currently doing that.

```javascript
var babelRelayPlugin = require('babel-relay-plugin');
var introspectionQuery = require('graphql/utilities').introspectionQuery;
var request = require('sync-request');

// TODO: Cache on filesystem
var response = request('GET', 'http://localhost:3000/graphql', {
  qs: {
    query: introspectionQuery
  }
});

var schema = JSON.parse(response.body.toString('utf-8'));

module.exports = babelRelayPlugin(schema.data);
```

**Comment by josephsavona** [comment]
@amccloud nice. a generic version of this that pulled configuration from `.graphql` would be pretty cool. PRs welcome! ;-)

**Comment by devknoll** [comment]
Yeah, this is basically the idea I was shooting for. A JSON-based `.graphqlrc` file that allowed you to specify a `schema`, that could either be a local path or a remote GraphQL endpoint.

It seems like it would also be very simple to make it handle a JS file exporting a graphql-js schema, and just have it run the same introspection query for that use case, right @josephsavona?

I'm still not entirely sure that it belongs in the main repo though... It would be nice for other tools if the `.graphqlrc` handling code was extracted into another module (but I suppose something like `require('babel-relay-plugin/graphqlrc')` wouldn't be that terrible...)

**Comment by josephsavona** [comment]
> It seems like it would also be very simple to make it handle a JS file exporting a graphql-js schema, and just have it run the same introspection query

Yup, we tried this but as I mentioned: "the schema file ends up require()-ing a different instance of graphql than the plugin does, and instanceof checks fail in GraphQLPrinter.js".

> I'm still not entirely sure that it belongs in the main repo though

I agree, but at the same time `babel-relay-plugin` already is a separate npm module (and might ultimately end up in a separate repo if that makes sense). For a new user experience, it would great if you could `npm install --save-dev babel-relay-plugin`, set up your `.graphqlrc` file, and have queries compile w/o any more effort. Regardless, I definitely agree that a config file and handler for various forms of schemas would be great to have.

**Comment by cesarandreu** [comment]
I think babel v6 will have a way to pass options in plugins, not 100% sure about the details though.

In any case, if you're adding support for a `.graphqlrc` file, please consider allowing it to be configured in your `package.json` as well. A lot of packages do this, as it lets you avoid having to keep adding files to your projects. (And having tons of config files that have to be kept in your project's root is annoying.)

**Comment by devknoll** [comment]
That's a good idea @cesarandreu. I wonder if there's an npm library that just takes care of all of this already...

**Comment by ForbesLindesay** [comment]
Using sync-request to fetch the schema from an endpoint inline within the babel plugin is a neat idea.  For working with the js-schema we could do a similar thing using spawn-sync (the underlying module that allows for synchronous requests).

**Comment by josephsavona** [comment]
So far we haven't heard any feedback about defining a `babelRelayPlugin.js` file as being a significant source of friction for users of Relay. I'm going to close this out - we encourage users to create wrappers around `babel-relay-plugin` to make it easy to define plugins that use local or remote schemas, using the suggestions listed here.

**Comment by F21** [comment]
I just came across this myself while setting up my dev environment to work with relay. While writing my own `baelRelayPlugin.js` isn't too onerous, I think it would be much nicer if I could do the following in my babel config (I am storing my bable config in `package.json`):

```json
"babel": {
    "presets": [
      "es2015",
      "stage-3",
      "react"
    ],
    "plugins": [
      ["babel-relay-plugin.js", {schema: "./schema.json"}]
    ]
  }
```

**Comment by BerndWessels** [comment]
Maybe unrelated but to use the `babel-relay-plugin` within the `.babelrc` you can use the [babel-plugin-loader](https://github.com/BerndWessels/babel-relay-plugin-loader) which lets you specify the location of the `schema.json` in your `package.json`.
Cheers
Bernd

**Comment by jinxac** [comment]
I was trying out something and I bumped into this.

I created first login of my `react-native` app using react-native and redux. 

So initially my `.babelrc` was

```
{
  "presets": [
    "react-native"
  ]
}
```
After this I installed my packages using `npm install`. 
Now my post login functionality is built up in graphql, so I thought lets bring in Relay, so I changed my .babelrc to following:-
```

{
  "passPerPreset": true,
  "presets": [
    {
      "plugins": [
        "./plugins/babelRelayPlugin"
      ]
    },
    "react-native"
  ]
}
```

Now, when i compiled my app again I was hitting the issue.

I deleted the `node_modules` folder and ran `npm install` again post which it works fine.
